### PR TITLE
Time Management Unite

### PR DESCRIPTION
**Pull request by Abdelbary**
Requesting merge to the master branch.

### CODE REVIEW

File: ServiceLayer/TMU/TMU.c [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+
+#include "TMU.h"
+
+#define TAKEN				1
+
+typedef struct TMU_obj_str
+{
+	/*object to hold data to the call back functions to call periodically*/
+	uint8_t Id;
+	void (*callB_fun)(void);
```
Comment (hossamadel) [inline ServiceLayer/TMU/TMU.c:16]: Use typedef to define the pointer to function

File: ServiceLayer/TMU/TMU.c [hunk 2]
```diff
@@ -0,0 +1,284 @@
...
+	uint8_t fire_tick;
+	uint8_t current_ticks;
+	uint8_t type;  /*periodic or one shot*/
+	
+}gstr_TMU_obj;
+
+static uint8_t TMU_Init_flag = FALSE;
+static uint8_t TMU_Timer_ch;
+static gstr_TMU_obj gastr_TMU_ObjBuffer[TMU_OBJ_BUFFER_SIZE];
```
Comment (hossamadel) [inline ServiceLayer/TMU/TMU.c:25]: Init the global variables inside the init function and deinit

File: ServiceLayer/TMU/TMU.c [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+		}
+		else if (tmu_cfg->tick_reslution > TMU_MAX_TIMER_RESLUTION 
+				|| tmu_cfg->timer_ch > TMU_TIMER_CH2)
+		{
+			fun_error_status = (TMU_MODULE_ERROR_NUM+INVALAD_PARAMETER);
+		}
+	}
+	else
+	{
+		TMU_Init_flag = TRUE;
```
Comment (hossamadel) [inline ServiceLayer/TMU/TMU.c:54]: Move this line at the end of the function

File: ServiceLayer/TMU/TMU.c [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+				timer_cfg.Timer_Polling_Or_Interrupt = TIMER_INTERRUPT_MODE;
+				timer_cfg.Timer_Prescaler = TIMER_PRESCALER_256;
+			break;
+			case TMU_TIMER_CH2:
+				timer_cfg.Timer_CH_NO = TIMER_CH2;
+				timer_cfg.Timer_Polling_Or_Interrupt = TIMER_INTERRUPT_MODE;
+				timer_cfg.Timer_Prescaler = TIMER_PRESCALER_256;
+			break;
+		}
+		Timer_Init(&timer_cfg);
```
Comment (hossamadel) [inline ServiceLayer/TMU/TMU.c:75]: Check the return value

File: ServiceLayer/TMU/TMU.c [hunk 5]
```diff
@@ -0,0 +1,284 @@
...
+	}
+	return fun_error_status;
+}
+ERROR_STATUS TMU_start(uint8_t Id,void (*callB_fun_ptr)(void),uint8_t lap_time,uint8_t type)
+{
+	sint16_t s16_fun_error_status = OK;
+	
+	if(Id > TMU_OBJ_BUFFER_SIZE || callB_fun_ptr == NULL || lap_time > TMU_MAX_LAP_TIME
```
Comment (hossamadel) [inline ServiceLayer/TMU/TMU.c:83]: More complicated condition

File: ServiceLayer/TMU/TMU.c [hunk 6]
```diff
@@ -0,0 +1,284 @@
...
+	}
+	else
+	{
+	static uint8_t fun_frstTime_flag = FALSE;
+	/*check for first time entrance to start timer*/
+	if(!fun_frstTime_flag)
+	{
+		/*start timer*/
+	
+		Timer_Start(TMU_Timer_ch,TIMER0_1MS_PRESCALER256);
```
Comment (hossamadel) [inline ServiceLayer/TMU/TMU.c:113]: Check the return value

File: ServiceLayer/TMU/TMU.c [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+		/*	check for new time tick
+		*	for each timer tick increase current tick and make timer0_MS_flag false
+		*	if current tick == lap_time
+		*		-zero current tick
+		*		-fire new_tick_flag
+		*/
+		if (timer0_MS_flag)
+		{
+			timer0_MS_flag = FALSE;
+			current_ticks++;
```
Comment (hossamadel) [inline ServiceLayer/TMU/TMU.c:204]: Increment the counter inside the ISR

File: ServiceLayer/TMU/TMU.c [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+	
+	if(TMU_Init_flag == FALSE)
+	{
+		/*report error*/
+		s16_fun_error_status = (TMU_MODULE_ERROR_NUM+MODULE_NOT_INITALIZED);
+	}
+
+	/*do deinit procedure anyway*/
+	uint8_t u8_counter = 0;
+	switch(TMU_Timer_ch)
```
Comment (hossamadel) [inline ServiceLayer/TMU/TMU.c:257]: Put the reset of the function in else condition

**Comment by hossamadel** [comment]
Thanks 😊

File: Test/TMU_test/TMU_test.c [hunk 9]
```diff
@@ -58,173 +61,287 @@ void stub_comsumer_fun(void)
...
 
+void checkTestCase(sint16_t s16_expected_return , sint16_t s16_fun_return )
+{
+	static uint8_t u8_testCaseNumber = 0;
+	printf("testCase %3d:	expected_return : %3d	 fun_return: %3d	",u8_testCaseNumber,s16_expected_return,s16_fun_return);
+	/*check on test case*/
+	if (s16_fun_return == s16_expected_return)
+	printf("Passed.\n");
+	else
+	printf("Failed.\n");
```
Comment (YehiaSprints) [inline Test/TMU_test/TMU_test.c:72]: It would be better to print the actual an the expected results for easy tracking and modifications.

**Pull request merged**